### PR TITLE
feat: make discovery field optional in plugins

### LINKED ISSUE #12027: Sidecar plugin requires discover command
Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

When using the plugin sidecar format, even when calling the plugin by name a `discover` command is still required.

**To Reproduce**

```
apiVersion: argoproj.io/v1alpha1
kind: ConfigManagementPlugin
metadata:
  name: my-plugin
spec:
  version: v1.0
  ...
  # no discover command
```

```
apiVersion: apiVersion: argoproj.io/v1alpha1
kind: ApplicationSet
metadata:
  name: testapp
spec:
  ...
  template:
    ...
    spec:
      project: test
      source:
        ...
        plugin:
          name: my-plugin-v1.0
          ...
```

Sidecar log: `invalid plugin configuration file. at least one of discover.find.name.command or discover.find.glob or discover.fineName should be non-empty`

**Expected behavior**

The appset runs through the plugin that has been specified by name, not having to put in a dummy discover command.

**Screenshots**

**Version**

```shell
v2.6.0-rc4
```

**Logs**

```
Paste any relevant application logs here.
```

### PR DESCRIPTION
**Pull request by gdsoumya**
Signed-off-by: Soumya Ghosh Dastidar <gdsoumya@gmail.com>

Fixes #12027 

This PR makes discovery field in cmpv2 plugin spec optional and removes discovery check for plugins when plugin name is explicitly specified in the app.

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

### CODE REVIEW

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]
I think this PR goes just a git too far. I think we should make the discovery field optional. But when it's present, I think it should have to match _in addition to_ the name field matching.

This will prevent 2.6 from introducing a new way for users to bypass discovery rules. This is important because Argo CD operators might be using discovery rules to enforce important patterns.

**Comment by gdsoumya** [comment]
> I think this PR goes just a git too far. I think we should make the discovery field optional. But when it's present, I think it should have to match _in addition to_ the name field matching.
> 
> This will prevent 2.6 from introducing a new way for users to bypass discovery rules. This is important because Argo CD operators might be using discovery rules to enforce important patterns.

There's no way to know if the discovery field is set or not so we just call the matchRepo rpc and try to check if the repo matches but if discovery is not specified it will automatically return false or no match for the request. I think that's probably fine because we do not break any existing implementations/patterns as previously discovery was compulsory so it's always there and users would probably not be putting plugin names in the app spec that won't work as the ui would have blocked it in the old version. WDYT?

**Comment by crenshaw-dev** [comment]
> we do not break any existing implementations/patterns as previously discovery was compulsory so it's always there

I'm worried specifically about the cases where `discovery` _is_ there, and the Argo CD operator expects it to continue to be enforced.

> if discovery is not specified it will automatically return false or no match for the request

I think instead of defaulting to false, we can default to true. If no discovery rules are set, `matchRepository` returns true.

**Comment by gdsoumya** [comment]
> > we do not break any existing implementations/patterns as previously discovery was compulsory so it's always there
> 
> I'm worried specifically about the cases where `discovery` _is_ there, and the Argo CD operator expects it to continue to be enforced.
> 
> > if discovery is not specified it will automatically return false or no match for the request
> 
> I think instead of defaulting to false, we can default to true. If no discovery rules are set, `matchRepository` returns true.

I see so the default behavior can be if discovery is not specified it will match all repos. That sounds good. Will make the change and update the docs to reflect the same.

**Comment by gdsoumya** [comment]
> I see so the default behavior can be if discovery is not specified it will match all repos. That sounds good. Will make the change and update the docs to reflect the same.

On second thought @crenshaw-dev do you think this would cause issues with auto-discovery? The reason I set it to not match by default was that I didn't want it to randomly match any repo that uses auto-discovery. I think that's more of a breaking case because older config using auto-discovery might see modified behavior after we enable this.

**Comment by gdsoumya** [comment]
> We would default to matching if and only if discovery is not set. As you mentioned earlier, every user of sidecar CMPs currently has discovery set, so it shouldn't be a breaking change. If discovery is set, we'll respect the discovery rules.

Yes but would that not break existing apps? for eg. if a user now creates a plugin with no discovery rules and because of how the plugin is named it get's more priority in the auto-discovery list then older apps that depended on auto-discovery would start matching with this plugin and because the plugin will always match with the repo the generated resources for those apps could be altered ?

**Comment by crenshaw-dev** [comment]
I understand the issue now. Good catch.

I think to satisfy all use cases, we need to modify the rpc so we can differentiate between "doesn't match because discovery" and "doesn't match because there _is_ no discovery".

File: cmpserver/plugin/plugin.go [hunk 1]
```diff
@@ -286,18 +286,19 @@ func (s *Service) matchRepositoryGeneric(stream MatchRepositoryStream) error {
 	return nil
 }
 
-func (s *Service) matchRepository(ctx context.Context, workdir string, envEntries []*apiclient.EnvEntry) (bool, error) {
+func (s *Service) matchRepository(ctx context.Context, workdir string, envEntries []*apiclient.EnvEntry) (bool, bool, error) {
```
Comment (crenshaw-dev) [inline cmpserver/plugin/plugin.go:289]: Can you add names to the return types for readability?

File: cmpserver/plugin/plugin.go [hunk 2]
```diff
@@ -286,18 +286,19 @@ func (s *Service) matchRepositoryGeneric(stream MatchRepositoryStream) error {
...
 	config := s.initConstants.PluginConfig
+
 	if config.Spec.Discover.FileName != "" {
```
Comment (crenshaw-dev) [inline cmpserver/plugin/plugin.go:292]: Do we now need a nil check before we start accessing `config.Spec.Discover` fields?

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]

File: cmpserver/plugin/plugin.go (hunk 2 above)
Comment (gdsoumya) [reply cmpserver/plugin/plugin.go:292]: No those are concrete structs and not pointers so nil check is not needed.

File: cmpserver/plugin/plugin.go (hunk 1 above)
Comment (gdsoumya) [reply cmpserver/plugin/plugin.go:289]: updated

**Review by crenshaw-dev — APPROVED** [review]

**Comment by crenshaw-dev** [comment]
Thanks, @gdsoumya!

**Comment by crenshaw-dev** [comment]
Cherry-picked onto release-2.6 for 2.6.0-rc8 or 2.6.0 GA, whichever comes first. :-)

**Pull request merged**